BUG: stabilize sort_values algorithms for Series and time-like Indices

AlexKirko opened the pull request:
- [X] closes #35922
- [X] 18 tests changed / 18 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

## Problem

Our `sort_values` functions currently behave differently for different objects: for most `Index` subclasses they are stable when sorting in descending order (this was introduced by #35604), but for DateTime-like `Index` subclasses and `Series` they are unstable. This isn't good as sorting should be stable across the board.

## Details

Came across this one while introducing missing-value support to `Index.sort_values` in #35604, so I had to limit that PR to non-DateTime-like `Index` subclasses. The problem was that we had different expectations for sorting stability baked into our test suite, so unifying sorting algorithms and missing-value support needed a bunch of careful test changes and altering both `sort_values` and algorithms in `sorting.py`.

Since this PR necessarily includes changes in several places, I have commented on all the changes made in the code and the unusual changes in the tests to make reviewing the code easier (see "On test changes" below).

## On test changes

Most changes I made in the tests are for cases where we were expecting an unstable sort or expected NaNs to be sorted to the beginning of a list of duplicates for ascending sort and to the end for descending (we forced this by inserting NaN-likes at 0 position and reversing when sorting in descending order in `Series.sort_values`).

## Default behavior changes

Since DateTime-like `Index` subclasses now support `na_position` using the same implementation as the other `Index` subclasses, they now sort missing values to the end of the `Index` by default.

## Performance

Ran the full benchmark suite, and there are no performance regressions.

## Out-of-scope

The only type I didn't touch so far is MultiIndex. It can't be sorted the same way through nargsort, and I don't think we should be doing it in this PR, if it all (stabilizing descending order MultiIndex.sort_values will definitely be a PITA, and it's a very narrow use case, in my opinion).

AlexKirko committed:
BUG: stabilize sorting in Series.sort_values

AlexKirko committed:
DOC: add comment to nargsort call in Series.sort_values

AlexKirko committed:
use nargsort with indices: Period, DateTime, TimeDelta

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko committed:
mv NaNs to the end of dupe lists in value_counts

pep8speaks commented:
Hello @AlexKirko! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/algorithms.py`](https://github.com/pandas-dev/pandas/blob/9b51d42f25b06cf06953c8be8235d55fc18aa4b5/pandas/core/algorithms.py):

> [Line 825:17](https://github.com/pandas-dev/pandas/blob/9b51d42f25b06cf06953c8be8235d55fc18aa4b5/pandas/core/algorithms.py#L825): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented (comment)
> [Line 826:17](https://github.com/pandas-dev/pandas/blob/9b51d42f25b06cf06953c8be8235d55fc18aa4b5/pandas/core/algorithms.py#L826): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented (comment)
> [Line 827:17](https://github.com/pandas-dev/pandas/blob/9b51d42f25b06cf06953c8be8235d55fc18aa4b5/pandas/core/algorithms.py#L827): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented (comment)

AlexKirko committed:
CLN: remove extra comment indents

AlexKirko committed:
attempt to mimic previous count_values behavior by reversing before sort

AlexKirko commented:
Stabilized the sorts, now looking for good ways to deal with 134 failing tests. Will likely take me a couple more days. Some tests have to be altered, but I'd like to change as few tests as necessary to get this done.

AlexKirko committed:
CLN: clean-up unnecessary import

AlexKirko committed:
Revert "attempt to mimic previous count_values behavior by reversing before sort"

This reverts commit 2b75f789d18840ec50d33eb83dae963e7f303c8c.

AlexKirko committed:
TST: alter tests in test_algos

AlexKirko commented on pandas/core/algorithms.py:
```diff
@@ -822,9 +822,19 @@ def value_counts_arraylike(values, dropna: bool):
+                keys[-1] = np.NaN
+                tmp = counts[nan_pos]
+                counts[nan_pos] = counts[-1]
+                counts[-1] = tmp
```
All of this is to ensure that NaNs appear after other items with duplicate counts when using value_counts. Could change the tests instead. I think that it's nice to have NaNs in a familiar spot, but it's not necessary.

AlexKirko replied:
Whether NaN ends up first or last or wherever among duplicates is largely irrelevant. Properly guaranteeing it being in the last position requires moving it to the end of `keys` and `counts` (recreating both arrays), and that's a performance cost with no significant upside. Better to just alter tests.

AlexKirko committed:
TST: alter value_counts dupe order in boolean/test_function

AlexKirko committed:
mv NaT to end of dupe sort order, alter test_value_counts

NaT is expected to be last in th elist of duplicate value counts
guarantee that by finding it and moving to the end of the array
(consider giving it up: the code ends up cluttered)
Also alter test_value_counts to make sure that expectations
match new sort_values stability

AlexKirko committed:
REFACT: use tuple unpacking for element swap

AlexKirko committed:
DOC: clarify comments in algorithms/value_counts

AlexKirko committed:
stop forcing NaN-like to be at the end of dupe order

AlexKirko committed:
TST: NaN-like is now first among duplicates in count_values

AlexKirko committed:
CLN: remove unnecessary is_bool import in series.py

AlexKirko committed:
TST: value_counts NaN dupe order change in test_string.py

AlexKirko committed:
TST: value_counts NaN dupe order in test_value_counts.py

AlexKirko committed:
CLN: rm unnecessary assignment from test_value_counts

AlexKirko committed:
TST: expect stable sort in extension/base/methods.py

AlexKirko committed:
BUG: support objs that raise when cast to their class

AlexKirko committed:
TST: fix stable sort expectation in test_sort_values in methods.py

AlexKirko committed:
TST: change top expect for dupe counts in frame/test_describe

AlexKirko committed:
BUG: clean up crutches in Series.nlargest

Now that sort_values is stable, we don't need to reverse
order in algorithms.SelectNSeries.compute, which is
the backend of Series.nlargest

AlexKirko committed:
TST: change dupe order expect in frame/methods/test_value_counts

AlexKirko committed:
TST: change dupe order expectation in indexes/datetimes/test_ops

AlexKirko committed:
TST: specify na-position in indexes/datetimes/test_ops.py

AlexKirko committed:
TST: specify na-position in indexes/period/test_ops.py

AlexKirko committed:
TST: remove xfail from test_order_stability_compat

AlexKirko committed:
TST: change dupe order expect for indexes/timedeltas/test_ops

AlexKirko committed:
BUG: reintroduce ascending param error-catching to sort_values

AlexKirko committed:
BUG: reintroduce proper key func support to Series.sort_values

AlexKirko committed:
BUG: fix bug in key func support

AlexKirko committed:
TST: alter dupe order expect in series/methods/test_value_counts

AlexKirko committed:
CLN: remove unused variable in Series.sort_values

AlexKirko committed:
BUG: set values in key func support in Series.sort_values

AlexKirko committed:
DOC: add whatsnew

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko committed:
BUG: add SparseArray sorting with key func support

AlexKirko committed:
TST: remove datetime-like xfails when testing indices with missing

AlexKirko committed:
TST: fix expect dupe sort order in doctests

AlexKirko committed:
BUG: fix bug in Series reconstruction after sorting

AlexKirko committed:
TST: fix expect dupe sort order in more doctests

AlexKirko committed:
BUG: support key func changing ndarray dtype

AlexKirko committed:
REFACT: reuse keyed

AlexKirko committed:
TST: remove datetime-like xfails from invalid_na_position

AlexKirko committed:
TST: fix expect dupe sort order in more doctests in base.py

AlexKirko commented on pandas/core/algorithms.py:
```diff
@@ -1181,10 +1181,8 @@ def compute(self, method: str) -> Series:
             ascending = method == "nsmallest"
-            slc = np.s_[::-1] if reverse_it else np.s_[:]
-            return dropped[slc].sort_values(ascending=ascending).head(n)
+            return dropped.sort_values(ascending=ascending).head(n)
```
Now that `sort_values` is stable, we don't need this extra reversal stuff for `nlargest` (it was used to stabilize the sort).

AlexKirko commented on pandas/core/indexes/base.py:
```diff
@@ -4525,9 +4525,7 @@ def sort_values(
-        if not isinstance(
-            self, (ABCMultiIndex, ABCDatetimeIndex, ABCTimedeltaIndex, ABCPeriodIndex)
-        ):
+        if not isinstance(self, ABCMultiIndex):
```
Support everything except `MultiIndex` (see PR header).

AlexKirko commented on pandas/core/series.py:
```diff
@@ -3307,21 +3284,24 @@ def _try_kind_sort(arr):
+        if key:
+            if isinstance(arr, SparseArray):
+                # SparseArray doesn't store NaNs item-by-item, so pass everything
+                arr = ensure_key_mapped(self, key)._values
```
SparseArrays can't be set by assignment, so just pass everything. Key func automatically runs on non-nans.

jreback replied:
hmm it would be nice to unify this

AlexKirko replied:
I agree, but so far have been unable to. Would appreciate any suggestions.

The problem  is that SparseArray doesn't support setting elements, so you can't do `arr[good] = something`. The workarounds I was able to Google are far too ugly to use.

jreback replied:
why do you need to convert this to `_values` beforet he assignment? i think that would make this agnostic (and just select ._valuses when passing nargsort

AlexKirko replied:
Thanks! Since we got rid of processing non-missing and missing values separately, this is now fixed.

AlexKirko commented on pandas/core/series.py:
```diff
@@ -3307,21 +3284,24 @@ def _try_kind_sort(arr):
+
+        result = self._constructor(
+            self._values[sorted_index], index=self.index[sorted_index]
+        )
```
Since we change `arr` when we sort with a key function, use original values here.

AlexKirko commented on pandas/core/series.py:
```diff
@@ -3307,21 +3284,24 @@ def _try_kind_sort(arr):
+                good = ~isna(arr)
+                keyed = ensure_key_mapped(self[good], key)._values
+                arr = arr.astype(keyed.dtype)
+                arr[good] = keyed
```
Nargsort expects its key function argument to take `ndarray` as arg, so we need to run the key func here, as `Series.sort_values` key function arg takes a Series.

AlexKirko commented on pandas/core/sorting.py:
```diff
@@ -499,7 +508,10 @@ def ensure_key_mapped(values, key: Optional[Callable], levels=None):
+            # GH 35922. Support sorting objects that raise when cast to their type
+            if not isinstance(result, type_of_values):
+                # try to revert to original type otherwise
+                result = type_of_values(result)
```
Some of the supported types don't like being cast to themselves.

jreback replied:
what does this fail on?

AlexKirko replied:
Seems like other fixes made this safeguard unnecessary. Now that we pass a `Series` into the function, there are no more failures. If I remember correctly, it was failing on tz-aware datetimes, when a key function would strip the timezone information, and then the cast would fail, because it wouldn't know what to do with the timezones in `values`.

AlexKirko commented on pandas/tests/extension/base/methods.py:
```diff
@@ -125,7 +125,11 @@ def test_sort_values(self, data_for_sorting, ascending, sort_by_key):
+            if ser.nunique() == 2:
+                expected = ser.iloc[[0, 1, 2]]
+            else:
+                expected = ser.iloc[[1, 0, 2]]
```
Some of our parametrizations in `data_for_sorting` contain dupliates and some don't. Since they are all structured the same way, we can detect different expected sort orders by checking the number of unique values. If it looks ugly to you, we could split `data_for_sorting` for this test.

jreback replied:
ok this is fine

AlexKirko commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -231,15 +231,15 @@ def test_order_without_freq(self, index_dates, expected_dates, tz_naive_fixture)
         expected = DatetimeIndex(expected_dates, tz=tz, name="idx")
 
-        ordered = index.sort_values()
+        ordered = index.sort_values(na_position="first")
```
NaTs used to be sorted to the beginning by default (because their numerical representation is -big_number). Now they get sorted to the end by default, and the user needs to supply the argument to get previous behavior. Wonder if this should be noted in the docs somewhere.

AlexKirko committed:
CLN: remove unnecessary imports in indexes/test_common.py

AlexKirko committed:
CLN: run black

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko commented:
@jreback 
This is ready for review, I think. Just finished running the benchmark suite, and everything is fine.

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -531,6 +531,7 @@ Other
 - Fixed metadata propagation in the :class:`Series.dt` and :class:`Series.str` accessors and :class:`DataFrame.duplicated` and ::class:`DataFrame.stack` methods (:issue:`28283`)
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is a :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError``, from a bare ``Exception`` previously (:issue:`35744`)
+- Sorting in descending order being unstable when using :meth:`Series.sort_values` and :meth:`Index.sort_values` for DateTime-like :class:`Index` subclasses (:issue:`35992`)
```
can you expand this slightly to say that the sorted results will be different than prior version, I think move this to the API changes section as well.

AlexKirko replied:
@jreback Sorry, I've expanded the message, but I can't seem to find an API changes section in 1.2.0 whatsnew. Could you point me to where it is in the file?

jreback commented on pandas/core/sorting.py:
```diff
@@ -374,7 +374,16 @@ def nargsort(
         non_nan_idx = non_nan_idx[::-1]
-    indexer = non_nan_idx[non_nans.argsort(kind=kind)]
+
+    # GH 35922. Move support for object sort here from Series.sort_values
```
this comment is not super useful, what is an expl of why you need this try/except e.g. L380 move here

AlexKirko replied:
No longer relevant, since I removed the try/except.

jreback commented on pandas/core/sorting.py:
```diff
@@ -374,7 +374,16 @@ def nargsort(
+
+    # GH 35922. Move support for object sort here from Series.sort_values
+    try:
+        # if kind==mergesort, it can fail for object dtype
```
why can this fail?

AlexKirko replied:
@jreback I moved this code from `Series.sort_values`, fearing that I don't understand why this is necessary, but it doesn't seem to be. I've removed the try/except, tested sorting `Series` of `object` with different `kind` settings, and ran the test suite, and nothing is going wrong. I think we can safely drop it.

jreback reviewed (changes requested):
cc @jbrockmendel if you'd have a look

jbrockmendel commented:
I'm not clear on what "stability" means in this context.  In #35922 you wrote

```
ind = pd.Index([1, 1, 2, 3])

>>> ind.sort_values(ascending=False, return_indexer=True)[1]
array([3, 2, 1, 0], dtype=int64)   # <-- old behavior
array([3, 2, 0, 1], dtype=int64)   # <-- new/expected behavior
```

Is it obvious why the latter is preferred?

AlexKirko committed:
incorporate both conflicting whatsnew

AlexKirko committed:
try removing if in ensure_key_mapped to find tests

AlexKirko commented:
> I'm not clear on what "stability" means in this context. In #35922 you wrote
> 
> ```
> ind = pd.Index([1, 1, 2, 3])
> 
> >>> ind.sort_values(ascending=False, return_indexer=True)[1]
> array([3, 2, 1, 0], dtype=int64)   # <-- old behavior
> array([3, 2, 0, 1], dtype=int64)   # <-- new/expected behavior
> ```
> 
> Is it obvious why the latter is preferred?

@jbrockmendel I think it's best to consider something like this right now on master:

```python
s = pd.Series(["A", "AA", "BB", "CAC"], dtype="object")
s.sort_values(ascending=False, key=lambda ser: ser.str.len())

OUT:
3    CAC
2     BB
1     AA
0      A
dtype: object
```
I don't think that swapping is expected here.

Then consider that you might be sorting a DataFrame with several columns, and a column with duplicates might be the first one. In this case you likely wouldn't expect a descending sort to change duplicate order. Or you could be using something like `nlargest` and get weirdness because there is a descending sort in there and it swaps elements.

Obviously we could get by with a convention that we always revert duplicate order with a descending sort by being careful, but I believe keeping duplicate order is cleaner. In cases where it doesn't matter, it's the same, and when it does matter (as in `nlargest` and the like), you don't need to remember that you need extra reversals.

AlexKirko committed:
BUG: just apply key func to the whole Series in sort_values

AlexKirko committed:
remove legacy try/except with default to quicksort

AlexKirko committed:
DOC: expand whatsnew

AlexKirko committed:
CLN: remove unnecessary SparseArray import

AlexKirko commented on pandas/core/series.py:
```diff
@@ -3307,21 +3284,16 @@ def _try_kind_sort(arr):
+            sorted_index = nargsort(ser._values, kind, ascending, na_position)
         else:
-            raise ValueError(f"invalid na_position: {na_position}")
+            sorted_index = nargsort(self._values, kind, ascending, na_position)
```
@jreback Thanks for the advice! I think that the separate handling of non-missing values and the whole `_values` stuff was necessary, because we were handling `na_position` here. But since we are now doing that in `nargsort`, we can just pass the whole Series to `ensure_key_mapped`, and be absolutely fine. This also gets rid of the need to handle `SparseArray` separately.

jreback replied:
great, i think can write like
```
values_to_sort = ensure_key_mapped(self, key)._values if key else self._values
sorted_index = nargsort(self._values, kind, ascending, na_position)
```

jreback replied:
leave your comment on L3290 as well

AlexKirko committed:
restart tests

AlexKirko committed:
restart tests again (windows_np18 misbehaving)

AlexKirko committed:
restart tests again

AlexKirko commented:
Not sure what's going on with the Window py37_np16 fail. Doesn't look like it has anything to do with this PR.

AlexKirko committed:
DOC: add information on sorting NaTs to whatsnew

AlexKirko committed:
DOC: phrasing change in whatsnew

jbrockmendel commented:
makes sense, thanks for spelling it out for me

jbrockmendel commented:
test_binary_arith_ops has been mucking up all the builds for a little while now; you can ignore it here

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko committed:
DOC: clarify whatsnew

AlexKirko committed:
CLN: clean up unnecessary newlines in sorting.py

Also bring back the cast comment

AlexKirko committed:
DOC: clarify whatsnew some more

AlexKirko committed:
bring back na_position validation in Series.sort_values

AlexKirko committed:
DOC: add to whatsnew

AlexKirko committed:
DOC: clarify NaTs sorting changes in whatsnew

AlexKirko committed:
DOC: add other api changes to whatsnew; move doc there

AlexKirko committed:
CLN: run black

AlexKirko commented:
@jreback 
I think this is ready for another look.

1. No longer process missing values separately in `Series.sort_values`, just pass everything to `ensure_key_mapping`.
2. Since we now pass the `Series` instead of the underlying array, we can now safely cast in `ensure_key_mapping`.
3. Clean up legacy try / except in nargsort (doesn't affect test suite, plus tested manually).
4. Answer @jbrockmendel 's question and add the answer to the issue for future reference.
5. Rewrite and expand whatsnew, create "Other API changes" section in whatsnew, and put it there.

More detailed answers to your comments in replies above.

The fails on `test_binary_arith_ops` are happening for all PRs and have nothing to do with this one.

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko committed:
DOC: attempt fixing malformed link in whatsnew

AlexKirko committed:
Revert "DOC: attempt fixing malformed link in whatsnew"

This reverts commit 36932cd5b4376b5f3459f5acc39b2e3d287ebf8c.

AlexKirko committed:
DOC: fix broken link in whatsnew

AlexKirko committed:
restart tests

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko commented:
@jreback 
Most of the unrelated stuff has been fixed. The Windows py38_np18 pipeline still fails, but that's unrelated and being examined in #37455 by @jbrockmendel

jreback reviewed (changes requested):
really a nice cleanup. just a small code comment, ping on green-ish.

AlexKirko committed:
REFACT: clean up key if/else in Series.sort_values

AlexKirko committed:
Merge branch 'master' into stable-dupe-sort

AlexKirko commented:
@jreback 
Made the change, all green-ish.

jreback commented:
thanks @AlexKirko very nice!

jorisvandenbossche commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -309,6 +309,13 @@ Optional libraries below the lowest tested version may still work, but are not c
+Other API changes
+^^^^^^^^^^^^^^^^^
+
+- Sorting in descending order is now stable for :meth:`Series.sort_values` and :meth:`Index.sort_values` for DateTime-like :class:`Index` subclasses. This will affect sort order when sorting :class:`DataFrame` on multiple columns, sorting with a key function that produces duplicates, or requesting the sorting index when using :meth:`Index.sort_values`. When using :meth:`Series.value_counts`, coun ...
```
Hey @AlexKirko, is it possible this is not only changed for datetime-like Index subclasses?

I see on released pandas this:

```
In [2]: pd.Index([0, 1, 0, 1]).value_counts()
Out[2]: 
1    2
0    2
dtype: int64
```

while on master / 1.2.0rc:

```
In [2]: pd.Index([0, 1, 0, 1]).value_counts()
Out[2]: 
0    2
1    2
dtype: int64
```

jorisvandenbossche replied:
Of course, it's the resulting Series that is sorted, not the Index. And I suppose for Series it's for all types?

(this change turned out to uncover a bug in statsmodels, https://github.com/statsmodels/statsmodels/pull/7215)
